plugins/last-fm/back.js

```diff
@@ -0,0 +1,192 @@
+	params.api_sig = api_sig;
+	for (key in params) {
+		queryData.push(`${key}=${params[key]}`);
+	}
+	return '?'+queryData.join('&');
+}
+
+const createApiSig = (params, secret) => { 
+	// this function creates the api signature, see: https://www.last.fm/api/authspec
+	let keys = [];
+	for (key in params) {
+		keys.push(key);
+	}
+	keys.sort();
+	let sig = '';
+	for (key of keys) {
+		if (String(key) === 'format')
+			continue
+		sig += `${key}${params[key]}`;
+	}
+	sig += secret;
+	sig = md5(sig);
+	return sig;
+}
+
+const createToken = async ({ api_key, api_root, secret }) => {
+	// creates and stores the auth token
+	data = {
+		method: 'auth.gettoken',
+		api_key: api_key,
+		format: 'json'
+	};
+	let api_sig = createApiSig(data, secret);
+	let response = await fetch(`${api_root}${createQueryString(data, api_sig)}`);
+	response = await response.json();
+	return response?.token;
+}
+
+const authenticate = async config => {
+	// asks the user for authentication
+	config.token = await createToken(config);
+	setOptions('last-fm', config);
+	open(`https://www.last.fm/api/auth/?api_key=${config.api_key}&token=${config.token}`);
+	return config;
+}
+
+const getAndSetSessionKey = async config => {
+	// get and store the session key
+	data = {
+		api_key: config.api_key,
+		format: 'json',
+		method: 'auth.getsession',
+		token: config.token,
+	};
+	api_sig = createApiSig(data, config.secret);
+	res = await fetch(`${config.api_root}${createQueryString(data, api_sig)}`);
+	res = await res.json();
+	if (res.error)
+		await authenticate(config);
+	config.session_key = res?.session?.key;
+	setOptions('last-fm', config);
+	return config;
+}
+
+
+const addScrobble = async (songInfo, config) => {
+	// this adds one scrobbled song to last.fm
+	if (!config.session_key)
+		await getAndSetSessionKey(config);
+	data = {
+		track: songInfo.title,
+		artist: songInfo.artist,
+		api_key: config.api_key,
+		sk: config.session_key,
+		format: 'json',
+		method: 'track.scrobble',
+		timestamp: ~~((Date.now() - songInfo.elapsedSeconds) / 1000),
+		duration: songInfo.songDuration,
+	};
+	data.api_sig = createApiSig(data, config.secret);
+	axios.post('https://ws.audioscrobbler.com/2.0/', createFormData(data))
+		.catch(res => {
+			if (res.response.data.error == 9) {
+				// session key is invalid, so remove it from the config and reauthenticate
+				config.session_key = undefined;
+				setOptions('last-fm', config);
+				authenticate(config);
+			}
+		});
+}
+
+const setNowPlaying = async (songInfo, config) => {
+	// this sets the now playing status in last.fm
+	if (!config.session_key)
+		await getAndSetSessionKey(config);
+	data = {
+		track: songInfo.title,
+		artist: songInfo.artist,
+		api_key: config.api_key,
+		sk: config.session_key,
+		format: 'json',
+		method: 'track.updateNowPlaying',
+		duration: songInfo.songDuration,
+	};
+	data.api_sig = createApiSig(data, config.secret);
+	axios.post('https://ws.audioscrobbler.com/2.0/', createFormData(data))
```

Comment:
Code looks mostly the same as `addScrobble` so it could probably be refactored with a util function.